torch/csrc/jit/passes/guard_elimination.cpp

```diff
@@ -0,0 +1,147 @@
+    // around prim::load, so we insert a dummy anchor to
+    // which we will be moving uses of params
+    // that alias analysis tells us are okay to move
+    auto start_node = b->owningGraph()->create(prim::Constant, 1);
+    start_node->output()->setType(IntType::create());
+    // the answer to the mystery of life
+    start_node->i_(attr::value, 42);
+    start_node->insertAfter(*b->nodes().begin());
+    for (auto it = b->nodes().begin(); it != b->nodes().end();) {
+      auto n = *it;
+      if (n->kind() == prim::Guard) {
+        // grab the next node before we move this one all the way back
+        it++;
+        auto guardee = n->inputs().at(0)->node();
+        if (guardee->kind() == prim::Param) {
+          guardee = start_node;
+        }
+        aliasDb_->moveAfterTopologicallyValid(n, guardee);
+      } else {
+        it++;
+        for (Block* ib : n->blocks()) {
+          moveGuardsToDefs(ib);
+        }
+      }
+    }
+  }
+
+  void coalesceGuards(Block* b) {
+    // uses on *all* parameters are moved to the same anchor node
+    // and they may come in different order after the anchor node
+    // e.g. (anchor, guard_x, guard_y, guard_x, guard_y)
+    // this pass recognizes contigious streches of guards and
+    // keeps track of the guards it's seen for each def. the next time
+    // the guard on the same def, it simply removes it.
+    std::unordered_map<Value*, Node*> inputs_to_guards;
+    for (auto it = b->nodes().begin(); it != b->nodes().end(); it++) {
+      auto n = *it;
+      if (n->kind() == prim::Guard) {
+        if (inputs_to_guards.count(n->input())) {
+          auto prev = inputs_to_guards[n->input()];
+          n->output()->replaceAllUsesWith(prev->output());
+          it.destroyCurrent();
+        } else {
+          inputs_to_guards.insert({n->input(), n});
+        }
+      } else if (n->kind() != prim::Constant) {
+        inputs_to_guards.clear();
+        for (Block* ib : n->blocks()) {
+          coalesceGuards(ib);
+        }
+      }
+    }
+  }
+
+  void eliminateGuards(Block* b) {
+    // a very simple pass to eliminate redundant guards for ops
+    // whose outputs are fully determined by their inputs
+    // i.e. if inputs to such ops are guarded we are allowed
+    // to remove a guard on ops' outputs
+    for (auto it = b->nodes().rbegin(); it != b->nodes().rend();) {
+      auto n = *it;
+      if (n->kind() == prim::Guard &&
+          removableGuard(n->inputs().at(0)->node())) {
+        auto pttp = n->output()->type();
+        n->output()->replaceAllUsesWith(n->inputs().at(0));
+        n->inputs().at(0)->setType(pttp);
+        it.destroyCurrent();
+      } else {
+        it++;
+        for (Block* ib : n->blocks()) {
+          eliminateGuards(ib);
+        }
+      }
+    }
+  }
+
+ private:
+  bool removableGuard(Node* n) {
+    if (!simple_ops_.count(n->kind())) {
+      return false;
+    }
+
+    bool all_inputs_guarded = true;
```

Comment:
Not sure this is the right structure for robust code longer term. It works for a list of simple_ops but for ops that have more complex behavior, we are going to have to have each op look at the profiled input type and output guard types and decide if the guard is redundant. It doesn't so much matter that the inputs to the op are guarded, moreso that they have a profiled tensor type that would make the output guard redundant. It is fine to merge this as is though, since we can revisit improving this later.